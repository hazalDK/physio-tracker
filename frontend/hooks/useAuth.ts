import * as SecureStore from "expo-secure-store";
import axios from "axios";
import { useNavigation } from "@react-navigation/native";
import { StackNavigationProp } from "@react-navigation/stack";
import { RootStackParamsList } from "@/types/navigation";
import { Alert } from "react-native";

export function useAuth() {
  const navigation =
    useNavigation<StackNavigationProp<RootStackParamsList, "login">>();

  const getToken = async () => {
    return await SecureStore.getItemAsync("access_token");
  };

  const refreshToken = async () => {
    try {
      const refreshToken = await SecureStore.getItemAsync("refresh_token");
      if (!refreshToken) {
        throw new Error("No refresh token available");
      }

      const refreshUrl = process.env.API_URL || "http://192.168.68.111:8000";
      const refreshResponse = await axios.post(
        `${refreshUrl}/api/token/refresh/`,
        { refresh: refreshToken }
      );

      const newToken = refreshResponse.data.access;
      const newRefreshToken = refreshResponse.data.refresh;

      await Promise.all([
        SecureStore.setItemAsync("access_token", newToken),
        SecureStore.setItemAsync("refresh_token", newRefreshToken),
      ]);

      return newToken;
    } catch (error) {
      await Promise.all([
        SecureStore.deleteItemAsync("access_token"),
        SecureStore.deleteItemAsync("refresh_token"),
      ]);

      Alert.alert("Session Expired", "Please login again", [
        { text: "OK", onPress: () => navigation.navigate("login") },
      ]);

      return null;
    }
  };

  const createApiInstance = async () => {
    let token = await getToken();

    if (!token) {
      Alert.alert("Login Required", "Please sign in to continue", [
        { text: "OK", onPress: () => navigation.navigate("login") },
      ]);
      return null;
    }

    return axios.create({
      baseURL: process.env.API_URL || "http://192.168.68.111:8000",
      timeout: 10000,
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });
  };

  return { getToken, refreshToken, createApiInstance };
}
