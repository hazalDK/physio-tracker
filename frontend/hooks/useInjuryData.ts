import { useEffect, useState } from "react";
import axios from "axios";
import { injuryTypesData } from "@/types/Injury";
import { getEnv } from "@/config";

/**
 * Custom hook to fetch and manage injury types data.
 * @returns {Object} - An object containing injury types data and loading state.
 */
export function useInjuryData(): {
  injuryTypes: injuryTypesData[];
  loading: boolean;
} {
  const [injuryTypes, setInjuryTypes] = useState<injuryTypesData[]>([]);
  const [loading, setLoading] = useState(true);

  // Function to fetch injury types data from the API and handle the loading state
  useEffect(() => {
    const fetchInjuryTypes = async () => {
      try {
        const apiUrl = getEnv("API_URL");
        const response = await axios.get<injuryTypesData[]>(
          `${apiUrl}/injury-types/`
        );
        setInjuryTypes(response.data);
      } catch (error) {
        console.error("Error fetching injury types:", error);
      } finally {
        setLoading(false);
      }
    };

    fetchInjuryTypes();
  }, []);

  return { injuryTypes, loading };
}
