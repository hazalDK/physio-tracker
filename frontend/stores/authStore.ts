// stores/authStore.ts
import { create } from "zustand";
import { persist, createJSONStorage } from "zustand/middleware";
import * as SecureStore from "expo-secure-store";

// Create an async storage adapter for SecureStore
const secureStorage = createJSONStorage(() => ({
  getItem: async (name: string) => {
    try {
      const value = await SecureStore.getItemAsync(name);
      return value ?? null;
    } catch (e) {
      console.error("Error getting item from SecureStore:", e);
      return null;
    }
  },
  setItem: async (name: string, value: string) => {
    try {
      await SecureStore.setItemAsync(name, value);
    } catch (e) {
      console.error("Error setting item in SecureStore:", e);
    }
  },
  removeItem: async (name: string) => {
    try {
      await SecureStore.deleteItemAsync(name);
    } catch (e) {
      console.error("Error removing item from SecureStore:", e);
    }
  },
}));

interface AuthStore {
  isAuthenticated: boolean;
  setIsAuthenticated: (value: boolean) => void;
  hydrated: boolean;
  setHydrated: (state: boolean) => void;
}

export const useAuthStore = create<AuthStore>()(
  persist(
    (set) => ({
      isAuthenticated: false,
      setIsAuthenticated: (value) => set({ isAuthenticated: value }),
      hydrated: false,
      setHydrated: (state) => set({ hydrated: state }),
    }),
    {
      name: "auth-storage",
      storage: secureStorage,
      // This is important - we'll handle hydration manually
      onRehydrateStorage: () => (state) => {
        if (state) {
          state.setHydrated(true);
        }
      },
    }
  )
);

// Add this function to manually trigger hydration in your app's entry point
export const hydrateAuthStore = async () => {
  const storedState = await SecureStore.getItemAsync("auth-storage");
  if (storedState) {
    try {
      const parsedState = JSON.parse(storedState);
      useAuthStore.setState({
        isAuthenticated: parsedState.state.isAuthenticated,
        hydrated: true,
      });
    } catch (e) {
      console.error("Failed to parse stored state:", e);
    }
  } else {
    useAuthStore.setState({ hydrated: true });
  }
};
