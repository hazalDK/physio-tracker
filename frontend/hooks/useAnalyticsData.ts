import { useState, useEffect } from "react";
import axios from "axios";
import { useAuth } from "./useAuth";

// Helper function to format date
const formatDate = (date: Date) => {
  return date.toISOString().split("T")[0];
};

// Type for analytics data (can be either adherence or pain data)
interface ChartData {
  labels: string[];
  datasets: {
    data: number[];
  }[];
}

/**
 * Custom hook to fetch and manage analytics data (adherence or pain levels).
 * @param {string} endpoint - The API endpoint to fetch data from.
 * @param {"adherence" | "pain"} dataType - The type of data to fetch (either adherence or pain).
 * @returns {Object} - An object containing loading state, chart data, average, history, error, and functions to navigate weeks.
 * @throws {Error} - Throws an error if the API call fails.
 * */
export function useAnalyticsData<T>(
  endpoint: string,
  dataType: "adherence" | "pain"
): object {
  const [loading, setLoading] = useState(true);
  const [chartData, setChartData] = useState<ChartData>({
    labels: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
    datasets: [{ data: [0, 0, 0, 0, 0, 0, 0] }],
  });
  const [average, setAverage] = useState(0);
  const [history, setHistory] = useState<T[]>([]);
  const [error, setError] = useState<string | null>(null);

  // Add week navigation state
  const [currentEndDate, setCurrentEndDate] = useState(new Date());
  const [weekTitle, setWeekTitle] = useState("Current Week");
  const { createApiInstance, refreshToken } = useAuth();

  // Function to go to previous week
  const goToPreviousWeek = () => {
    const newEndDate = new Date(currentEndDate);
    newEndDate.setDate(newEndDate.getDate() - 7);
    setCurrentEndDate(newEndDate);
    fetchData(newEndDate);
    updateWeekTitle(newEndDate);
  };

  // Function to go to next week (but not beyond current date)
  const goToNextWeek = () => {
    const today = new Date();
    const newEndDate = new Date(currentEndDate);
    newEndDate.setDate(newEndDate.getDate() + 7);

    // Don't go beyond current date
    if (newEndDate > today) {
      newEndDate.setTime(today.getTime());
    }

    setCurrentEndDate(newEndDate);
    fetchData(newEndDate);
    updateWeekTitle(newEndDate);
  };

  // Update week title based on date range
  const updateWeekTitle = (endDate: Date) => {
    const startDate = new Date(endDate);
    startDate.setDate(startDate.getDate() - 6);

    const today = new Date();
    const isCurrentWeek =
      today.getDate() === endDate.getDate() &&
      today.getMonth() === endDate.getMonth() &&
      today.getFullYear() === endDate.getFullYear();

    if (isCurrentWeek) {
      setWeekTitle("Current Week");
    } else {
      const startMonth = startDate.toLocaleString("default", {
        month: "short",
      });
      const endMonth = endDate.toLocaleString("default", { month: "short" });

      if (startMonth === endMonth) {
        setWeekTitle(
          `${startMonth} ${startDate.getDate()} - ${endDate.getDate()}`
        );
      } else {
        setWeekTitle(
          `${startMonth} ${startDate.getDate()} - ${endMonth} ${endDate.getDate()}`
        );
      }
    }
  };

  // Function to fetch data for a specific end date
  const fetchData = async (endDate: Date) => {
    setLoading(true);
    try {
      const api = await createApiInstance();
      if (!api) return;

      try {
        // Fetch stats from the API with date parameter
        const response = await api.get(
          `${endpoint}?end_date=${formatDate(endDate)}`
        );
        if (response.data) {
          setChartData(response.data.chart_data);

          if (dataType === "adherence") {
            setAverage(response.data.average_adherence);
          } else {
            setAverage(response.data.average_pain);
          }

          setHistory(response.data.history);
        }
      } catch (error) {
        // Handle 401 error by refreshing the token and retrying the request
        if (axios.isAxiosError(error) && error.response?.status === 401) {
          // Attempt token refresh
          const newToken = await refreshToken();
          if (!newToken) return;

          const api = await createApiInstance();
          if (!api) return;

          // Retry with new token
          const response = await api.get(
            `${endpoint}?end_date=${formatDate(endDate)}`
          );

          if (response.data) {
            setChartData(response.data.chart_data);

            if (dataType === "adherence") {
              setAverage(response.data.average_adherence);
            } else {
              setAverage(response.data.average_pain);
            }

            setHistory(response.data.history);
          }
        } else {
          // Handle other errors
          console.error("API request failed:", error);
          setError(
            `Failed to load your ${
              dataType === "adherence" ? "exercises" : "pain data"
            }. Please check your connection and try again.`
          );
        }
      }
    } catch (error) {
      console.error(
        `Failed to load your ${
          dataType === "adherence" ? "exercises" : "pain data"
        }.`
      );
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    // Initial data fetch for current week
    fetchData(currentEndDate);
    updateWeekTitle(currentEndDate);
  }, []);

  return {
    loading,
    chartData,
    average,
    history,
    error,
    weekTitle,
    goToPreviousWeek,
    goToNextWeek,
  };
}
