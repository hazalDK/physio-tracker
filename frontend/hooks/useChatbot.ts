import { useState, useEffect, useCallback } from "react";
import { Alert } from "react-native";
import Toast from "react-native-root-toast";
import { messageType } from "@/types/chatbot";
import { createChatApi, apiRequestWithRefresh } from "../api";
import { useAuth } from "./useAuth";

export default function useChatbot() {
  const auth = useAuth();
  const [messages, setMessages] = useState<messageType[]>([]);
  const [userExercises, setUserExercises] = useState([]);
  const [recentPain, setRecentPain] = useState(0);
  const [isLoading, setIsLoading] = useState(false);
  const [isInitialLoading, setIsInitialLoading] = useState(true);
  const [api, setApi] = useState<any>(null);

  // Initialize API and chat
  useEffect(() => {
    const initializeApi = async () => {
      const apiInstance = await auth.createApiInstance();
      setApi(apiInstance);
    };

    initializeApi();

    // Set initial messages
    setMessages([
      {
        id: "disclaimer",
        text: "DISCLAIMER",
        sender: "disclaimer",
      },
      {
        id: "welcome",
        text: "Welcome to your physiotherapy assistant! How can I help you today?",
        sender: "bot",
      },
    ]);
  }, []);

  // Fetch user data when API is available
  useEffect(() => {
    if (api) {
      fetchUserData();
    }
  }, [api]);

  const fetchUserData = async () => {
    if (!api) return;

    setIsInitialLoading(true);
    const chatApi = createChatApi(api);

    try {
      // Use the wrapper function for API calls with refresh
      const fetchData = async () => {
        const [reportsData, exercisesData] = await Promise.all([
          chatApi.fetchReports(),
          chatApi.fetchUserExercises(),
        ]);

        return { reportsData, exercisesData };
      };

      const { reportsData, exercisesData } = await apiRequestWithRefresh(
        fetchData,
        auth
      );

      setUserExercises(exercisesData);

      if (reportsData.length > 0) {
        setRecentPain(reportsData[0].pain_level);
      }
    } catch (error) {
      console.error("Failed to fetch user data:", error);

      Alert.alert(
        "Error",
        "Failed to load your exercises. Please check your connection and try again."
      );
    } finally {
      setIsInitialLoading(false);
    }
  };

  const sendMessage = async (inputText: string) => {
    if (!inputText.trim() || !api) return;

    // Add user message to chat
    const userMessageId = Date.now().toString();
    setMessages((prev) => [
      ...prev,
      { id: userMessageId, text: inputText, sender: "user" },
    ]);

    setIsLoading(true);
    const chatApi = createChatApi(api);

    try {
      // Create context object with user's exercise data
      const exerciseContext = {
        exercises: userExercises,
        recentPain: recentPain,
      };

      // Use the wrapper function for API calls with refresh
      const sendMessageCall = async () => {
        return await chatApi.sendChatMessage(inputText, exerciseContext);
      };

      // Send message to API with refresh handling
      const response = await apiRequestWithRefresh(sendMessageCall, auth);

      // Add bot response to chat
      setMessages((prev) => [
        ...prev,
        {
          id: Date.now().toString(),
          text: response.message,
          sender: "bot",
        },
      ]);
    } catch (error) {
      console.error("Error sending message:", error);

      setMessages((prev) => [
        ...prev,
        {
          id: Date.now().toString(),
          text: "Sorry, I encountered an error. Please try again.",
          sender: "bot",
        },
      ]);
    } finally {
      setIsLoading(false);
    }
  };

  const resetChat = async () => {
    if (!api) return;

    const chatApi = createChatApi(api);

    try {
      // Use the wrapper function for API calls with refresh
      const resetCall = async () => {
        await chatApi.resetChat();
      };

      await apiRequestWithRefresh(resetCall, auth);

      // Reset messages
      setMessages([
        {
          id: "disclaimer",
          text: "DISCLAIMER",
          sender: "disclaimer",
        },
        {
          id: "welcome",
          text: "Welcome to your physiotherapy assistant! How can I help you today?",
          sender: "bot",
        },
      ]);

      // Refresh user data
      fetchUserData();

      Toast.show("Chat reset!", {
        duration: Toast.durations.SHORT,
        position: Toast.positions.BOTTOM,
        backgroundColor: "#14b8a6",
        textColor: "white",
        shadow: true,
      });
    } catch (error) {
      console.error("Failed to reset chat:", error);
      Alert.alert("Error", "Could not reset chat.");
    }
  };

  return {
    messages,
    isLoading,
    isInitialLoading,
    sendMessage,
    resetChat,
  };
}
