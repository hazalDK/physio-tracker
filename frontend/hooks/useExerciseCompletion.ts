import { useState } from "react";
import { Alert } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { useAuth } from "./useAuth";

// Custom hook to manage user profile data and related operations.
export function useExerciseCompletion(userExerciseId: number | null) {
  const [showCompletionForm, setShowCompletionForm] = useState(false);
  const [reps, setReps] = useState(0);
  const [sets, setSets] = useState(0);
  const [painLevel, setPainLevel] = useState(0);
  const navigation = useNavigation();
  const { createApiInstance } = useAuth();

  // Function to handle saving exercise completion data
  // This function is called when the user submits the completion form.
  const handleSave = async () => {
    // Validate input fields
    if (!reps || !sets || !painLevel) {
      Alert.alert("Error", "Please fill all required fields");
      return;
    }

    // Api instance creation
    const api = await createApiInstance();
    if (!api) return;

    try {
      // Update user exercise data
      await api.put(`/user-exercises/${userExerciseId}/`, {
        reps: reps,
        sets: sets,
        pain_level: painLevel,
        completed: true,
      });

      // Closes the modal
      setShowCompletionForm(false);

      // Handle pain level logic
      // If pain level is low (1-3), check if they can increase difficulty
      if (painLevel <= 3) {
        try {
          const canIncreaseResponse = await api.get(
            `/user-exercises/${userExerciseId}/can_increase/`
          );

          if (canIncreaseResponse.data.can_increase) {
            Alert.alert(
              "Exercise Progress",
              "Your pain level is low and you've had consistent low pain for 3 days. Would you like to increase the difficulty for next time?",
              [
                {
                  text: "No",
                  onPress: () => navigation.goBack(),
                  style: "cancel",
                },
                {
                  text: "Yes",
                  onPress: async () => {
                    try {
                      await api.post(
                        `/user-exercises/${userExerciseId}/confirm_increase/`,
                        { confirm: "yes" }
                      );
                      Alert.alert(
                        "Success",
                        "Exercise difficulty increased for next time!"
                      );
                      navigation.goBack();
                    } catch (error) {
                      console.error("Error increasing difficulty:", error);
                      Alert.alert(
                        "Error",
                        "Failed to increase difficulty. Please try again."
                      );
                      navigation.goBack();
                    }
                  },
                },
              ]
            );
          } else {
            Alert.alert(
              "Exercise Progress",
              "Your pain level is low. Keep it up for 3 days to unlock higher difficulty!",
              [{ text: "OK", onPress: () => navigation.goBack() }]
            );
          }
        } catch (error) {
          // Handle error when checking increase eligibility
          console.error("Error checking increase eligibility:", error);
          Alert.alert("Success", "Exercise completed successfully", [
            { text: "OK", onPress: () => navigation.goBack() },
          ]);
        }
      } else if (painLevel > 3) {
        // If pain level is high, ask if they want to decrease difficulty
        Alert.alert(
          "Exercise Difficulty",
          "Your pain level is high. Would you like to decrease the difficulty for next time?",
          [
            {
              text: "No",
              onPress: () => navigation.goBack(),
              style: "cancel",
            },
            {
              text: "Yes",
              onPress: async () => {
                try {
                  await api.post(
                    `/user-exercises/${userExerciseId}/confirm_decrease/`,
                    { confirm: "yes" }
                  );
                  Alert.alert(
                    "Success",
                    "Exercise difficulty decreased for next time!"
                  );
                  navigation.goBack();
                } catch (error) {
                  console.error("Error decreasing difficulty:", error);
                  Alert.alert(
                    "Error",
                    "Failed to decrease difficulty. Please try again."
                  );
                  navigation.goBack();
                }
              },
            },
          ]
        );
      } else {
        Alert.alert("Success", "Exercise completed successfully", [
          { text: "OK", onPress: () => navigation.goBack() },
        ]);
      }
    } catch (error) {
      // Handle error when saving exercise completion
      console.error("Error saving exercise:", error);
      Alert.alert(
        "Error",
        "Failed to save exercise completion. Please try again."
      );
    }
  };

  // Generate dropdown data
  const getDropdownData = (maxValue: number) => {
    return Array.from({ length: maxValue }, (_, i) => ({
      label: `${i + 1}`,
      value: i + 1,
    }));
  };

  // Generate pain level dropdown data
  const getPainLevelData = () => {
    return Array.from({ length: 11 }, (_, i) => ({
      label: `${i}`,
      value: i,
    }));
  };

  return {
    showCompletionForm,
    setShowCompletionForm,
    reps,
    setReps,
    sets,
    setSets,
    painLevel,
    setPainLevel,
    handleSave,
    getDropdownData,
    getPainLevelData,
  };
}
