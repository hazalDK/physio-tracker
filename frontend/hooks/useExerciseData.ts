import { useState, useEffect } from "react";
import { Alert } from "react-native";
import axios from "axios";
import { ExerciseItem, UserExerciseItem } from "@/types/exercise";
import { useAuth } from "./useAuth";

export function useExerciseData(
  exerciseId: number,
  userExerciseId: number | null
) {
  const [loading, setLoading] = useState(true);
  const [exercise, setExercise] = useState<ExerciseItem>();
  const [userExercise, setUserExercise] = useState<UserExerciseItem>();
  const { createApiInstance, refreshToken } = useAuth();

  useEffect(() => {
    const fetchData = async () => {
      setLoading(true);
      try {
        const api = await createApiInstance();
        if (!api) return;

        try {
          const [userExercisesResponse, exercisesResponse] = await Promise.all([
            api.get(`/user-exercises/${userExerciseId}/`),
            api.get(`/exercises/${exerciseId}/`),
          ]);

          setUserExercise(
            !Array.isArray(userExercisesResponse.data)
              ? userExercisesResponse.data
              : ({} as UserExerciseItem)
          );

          setExercise(
            !Array.isArray(exercisesResponse.data)
              ? exercisesResponse.data
              : ({} as ExerciseItem)
          );
        } catch (error) {
          if (axios.isAxiosError(error) && error.response?.status === 401) {
            const newToken = await refreshToken();
            if (!newToken) return;

            const api = await createApiInstance();
            if (!api) return;

            const [retryUserExercises, retryExercises] = await Promise.all([
              api.get(`/user-exercises/${userExerciseId}/`),
              api.get(`/exercises/${exerciseId}/`),
            ]);

            setUserExercise(
              !Array.isArray(retryUserExercises.data)
                ? retryUserExercises.data
                : ({} as UserExerciseItem)
            );

            setExercise(
              !Array.isArray(retryExercises.data)
                ? retryExercises.data
                : ({} as ExerciseItem)
            );
          } else {
            console.error("API request failed:", error);
            Alert.alert(
              "Error",
              "Failed to load your exercises. Please check your connection and try again."
            );
            setExercise({} as ExerciseItem);
            setUserExercise({} as UserExerciseItem);
          }
        }
      } catch (error) {
        console.error("Failed to load exercise data:", error);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [exerciseId, userExerciseId]);

  return { loading, exercise, userExercise };
}
