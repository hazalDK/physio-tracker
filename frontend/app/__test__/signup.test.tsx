import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { Alert } from "react-native";
import renderer from "react-test-renderer";
import axios from "axios";
import { NavigationContainer } from "@react-navigation/native";
import * as SecureStore from "expo-secure-store";
import { router } from "expo-router";
import Signup from "../signup";
import { useInjuryData } from "@/hooks/useInjuryData";
import { useAuthStore } from "@/stores/authStore";

// Mock dependencies
jest.mock("axios");
jest.mock("expo-secure-store", () => ({
  setItemAsync: jest.fn(() => Promise.resolve()),
  getItemAsync: jest.fn(() => Promise.resolve(null)),
}));
jest.mock("expo-router", () => ({
  router: {
    replace: jest.fn(),
  },
}));
jest.mock("@react-navigation/native", () => ({
  Link: ({ children }) => children,
  NavigationContainer: ({ children }) => children,
}));
jest.mock("react-native", () => {
  const RN = jest.requireActual("react-native");
  RN.Alert.alert = jest.fn();
  return RN;
});
jest.mock("react-native-gesture-handler", () => ({
  TextInput: "TextInput",
  TouchableOpacity: "TouchableOpacity",
  ScrollView: "ScrollView",
}));
jest.mock("react-native-element-dropdown", () => ({
  Dropdown: "Dropdown",
}));
jest.mock("@react-native-community/datetimepicker", () => "DateTimePicker");
jest.mock("tailwind-react-native-classnames", () => ({
  __esModule: true,
  default: () => ({}),
}));
jest.mock("@/hooks/useInjuryData", () => ({
  useInjuryData: jest.fn(),
}));
jest.mock("@/stores/authStore", () => ({
  useAuthStore: jest.fn(),
}));
jest.mock("react-native-safe-area-context", () => ({
  useSafeAreaInsets: () => ({ top: 0, bottom: 0, left: 0, right: 0 }),
  SafeAreaProvider: ({ children }) => children,
}));

// Use the actual Signup component instead of mocking it
jest.unmock("../signup");

// Create a wrapper component with NavigationContainer
const SignupWithNavigation = (props) => (
  <NavigationContainer>
    <Signup {...props} />
  </NavigationContainer>
);

describe("Signup Component", () => {
  const mockSetIsAuthenticated = jest.fn();
  let alertMock: jest.SpyInstance;
  const mockSetInjuryType = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    alertMock = jest.spyOn(Alert, "alert");

    // Mock the useInjuryData hook
    (useInjuryData as jest.Mock).mockReturnValue({
      injuryTypes: [
        { id: 1, name: "Fracture" },
        { id: 2, name: "Sprain" },
      ],
      loading: false,
    });

    // Mock the authStore
    (useAuthStore as unknown as jest.Mock).mockImplementation((callback) => {
      return callback({ setIsAuthenticated: mockSetIsAuthenticated });
    });

    // Set default environment variable
    process.env.API_URL = "http://test-api.com";

    // Mock console.error to suppress expected error logs
    jest.spyOn(console, "error").mockImplementation(() => {});
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("renders correctly with data", () => {
    const tree = renderer.create(<SignupWithNavigation />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders all form fields correctly", () => {
    const { getByText, getByPlaceholderText } = render(
      <SignupWithNavigation
        testInjuryType={undefined}
        setInjuryTypeForTest={undefined}
      />
    );

    expect(getByText("Registration")).toBeTruthy();
    expect(getByPlaceholderText("Enter first name here")).toBeTruthy();
    expect(getByPlaceholderText("Enter last name here")).toBeTruthy();
    expect(getByPlaceholderText("Enter username here")).toBeTruthy();
    expect(getByPlaceholderText("example@gmail.com")).toBeTruthy();
    expect(getByText("Date of Birth")).toBeTruthy();
    expect(getByText("Injury Type")).toBeTruthy();
  });

  it("handles input changes for all fields", () => {
    const { getByPlaceholderText } = render(<SignupWithNavigation />);

    const firstNameInput = getByPlaceholderText("Enter first name here");
    const lastNameInput = getByPlaceholderText("Enter last name here");
    const usernameInput = getByPlaceholderText("Enter username here");
    const emailInput = getByPlaceholderText("example@gmail.com");

    fireEvent.changeText(firstNameInput, "John");
    fireEvent.changeText(lastNameInput, "Doe");
    fireEvent.changeText(usernameInput, "johndoe");
    fireEvent.changeText(emailInput, "john@example.com");

    expect(firstNameInput.props.value).toBe("John");
    expect(lastNameInput.props.value).toBe("Doe");
    expect(usernameInput.props.value).toBe("johndoe");
    expect(emailInput.props.value).toBe("john@example.com");
  });

  it("validates email format and shows error message for invalid email", () => {
    const { getByPlaceholderText, getByText } = render(
      <SignupWithNavigation />
    );

    const emailInput = getByPlaceholderText("example@gmail.com");

    // Input invalid email
    fireEvent.changeText(emailInput, "invalid-email");

    // Check for error message
    expect(
      getByText("Please enter a valid email (e.g., user@example.com).")
    ).toBeTruthy();

    // Input valid email should clear error
    fireEvent.changeText(emailInput, "valid@example.com");
    expect(() =>
      getByText("Please enter a valid email (e.g., user@example.com).")
    ).toThrow();
  });

  it("validates password requirements and shows error message", () => {
    const { getByTestId, getByText } = render(<SignupWithNavigation />);

    const passwordInput = getByTestId("password-input");

    // Input invalid password (too short)
    fireEvent.changeText(passwordInput, "short");

    // Check for error message
    expect(
      getByText(
        "Password must be 8+ chars with uppercase, lowercase, number, and special char."
      )
    ).toBeTruthy();

    // Input valid password should clear error
    fireEvent.changeText(passwordInput, "ValidP@ss1");
    expect(() =>
      getByText(
        "Password must be 8+ chars with uppercase, lowercase, number, and special char."
      )
    ).toThrow();
  });

  it("shows an alert when form is submitted with missing fields", async () => {
    const { getByText } = render(
      <SignupWithNavigation
        testInjuryType={1}
        setInjuryTypeForTest={mockSetInjuryType}
      />
    );

    // Submit without filling any fields
    fireEvent.press(getByText("Sign Up"));

    await waitFor(() => {
      expect(alertMock).toHaveBeenCalledWith(
        "Error",
        "Please fill all required fields"
      );
    });
  });

  it("shows an alert when passwords do not match", async () => {
    // Render with the test injury type prop
    const { getByPlaceholderText, getByText, getByTestId } = render(
      <SignupWithNavigation
        testInjuryType={1}
        setInjuryTypeForTest={undefined}
      />
    );

    // Fill all required fields
    fireEvent.changeText(getByPlaceholderText("Enter first name here"), "John");
    fireEvent.changeText(getByPlaceholderText("Enter last name here"), "Doe");
    fireEvent.changeText(
      getByPlaceholderText("Enter username here"),
      "johndoe"
    );
    fireEvent.changeText(
      getByPlaceholderText("example@gmail.com"),
      "john@example.com"
    );

    // Fill passwords with different values
    const passwordInput = getByTestId("password-input");
    fireEvent.changeText(passwordInput, "Password1!");

    const confirmPasswordInput = getByTestId("confirm-password-input");
    fireEvent.changeText(confirmPasswordInput, "DifferentPass1!");

    // Submit the form
    fireEvent.press(getByText("Sign Up"));

    await waitFor(() => {
      expect(alertMock).toHaveBeenCalledWith("Error", "Passwords do not match");
    });
  });

  it("handles successful registration", async () => {
    // Mock axios post to return successful response
    (axios.post as jest.Mock).mockResolvedValueOnce({
      data: {
        access: "fake-access-token",
        refresh: "fake-refresh-token",
      },
    });

    // Render with the test injury type prop
    const { getByPlaceholderText, getByText, getByTestId } = render(
      <SignupWithNavigation
        testInjuryType={1}
        setInjuryTypeForTest={undefined}
      />
    );

    // Fill out the form with valid data
    fireEvent.changeText(getByPlaceholderText("Enter first name here"), "John");
    fireEvent.changeText(getByPlaceholderText("Enter last name here"), "Doe");
    fireEvent.changeText(
      getByPlaceholderText("Enter username here"),
      "johndoe"
    );
    fireEvent.changeText(
      getByPlaceholderText("example@gmail.com"),
      "john@example.com"
    );

    // Set matching passwords
    const passwordInput = getByTestId("password-input");
    fireEvent.changeText(passwordInput, "Password1!");

    const confirmPasswordInput = getByTestId("confirm-password-input");
    fireEvent.changeText(confirmPasswordInput, "Password1!");

    // Submit the form
    fireEvent.press(getByText("Sign Up"));

    await waitFor(() => {
      // Verify API was called with correct data
      expect(axios.post).toHaveBeenCalledWith(
        "http://test-api.com/users/register/",
        expect.objectContaining({
          username: "johndoe",
          email: "john@example.com",
          password: "Password1!",
          first_name: "John",
          last_name: "Doe",
          injury_type: 1,
        }),
        expect.any(Object)
      );

      // Verify that SecureStore was called to store tokens
      expect(SecureStore.setItemAsync).toHaveBeenCalledWith(
        "access_token",
        "fake-access-token"
      );
      expect(SecureStore.setItemAsync).toHaveBeenCalledWith(
        "refresh_token",
        "fake-refresh-token"
      );

      // Verify auth state was updated
      expect(mockSetIsAuthenticated).toHaveBeenCalledWith(true);

      // Verify that router was called to navigate to home
      expect(router.replace).toHaveBeenCalledWith("/(tabs)");
    });
  });

  it("handles registration failure with server errors - username", async () => {
    // Mock axios post to return an error with field errors
    (axios.post as jest.Mock).mockRejectedValueOnce({
      response: {
        data: {
          username: ["This username is already taken."],
        },
      },
    });

    // Render with the test injury type prop
    const { getByPlaceholderText, getByText, getByTestId } = render(
      <SignupWithNavigation
        testInjuryType={1}
        setInjuryTypeForTest={undefined}
      />
    );

    // Fill out the form with valid data
    fireEvent.changeText(getByPlaceholderText("Enter first name here"), "John");
    fireEvent.changeText(getByPlaceholderText("Enter last name here"), "Doe");
    fireEvent.changeText(
      getByPlaceholderText("Enter username here"),
      "johndoe"
    );
    fireEvent.changeText(
      getByPlaceholderText("example@gmail.com"),
      "john@example.com"
    );

    // Set matching passwords
    const passwordInput = getByTestId("password-input");
    fireEvent.changeText(passwordInput, "Password1!");

    const confirmPasswordInput = getByTestId("confirm-password-input");
    fireEvent.changeText(confirmPasswordInput, "Password1!");

    // Submit the form
    fireEvent.press(getByText("Sign Up"));

    await waitFor(() => {
      expect(alertMock).toHaveBeenCalledWith(
        "Registration Failed",
        "Username: This username is already taken."
      );
    });
  });

  it("handles registration failure with server errors - email", async () => {
    // Mock axios post to return an error with email field errors
    (axios.post as jest.Mock).mockRejectedValueOnce({
      response: {
        data: {
          email: ["A user with this email already exists."],
        },
      },
    });

    // Render with the test injury type prop
    const { getByPlaceholderText, getByText, getByTestId } = render(
      <SignupWithNavigation
        testInjuryType={1}
        setInjuryTypeForTest={undefined}
      />
    );

    // Fill out the form with valid data
    fireEvent.changeText(getByPlaceholderText("Enter first name here"), "John");
    fireEvent.changeText(getByPlaceholderText("Enter last name here"), "Doe");
    fireEvent.changeText(
      getByPlaceholderText("Enter username here"),
      "johndoe"
    );
    fireEvent.changeText(
      getByPlaceholderText("example@gmail.com"),
      "john@example.com"
    );

    // Set matching passwords
    const passwordInput = getByTestId("password-input");
    fireEvent.changeText(passwordInput, "Password1!");

    const confirmPasswordInput = getByTestId("confirm-password-input");
    fireEvent.changeText(confirmPasswordInput, "Password1!");

    // Submit the form
    fireEvent.press(getByText("Sign Up"));

    await waitFor(() => {
      expect(alertMock).toHaveBeenCalledWith(
        "Registration Failed",
        "Email: A user with this email already exists."
      );
    });
  });
});
