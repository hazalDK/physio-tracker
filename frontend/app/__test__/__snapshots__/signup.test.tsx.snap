// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Component renders correctly with data 1`] = `null`;
